clippy_lints/src/use_self.rs

```diff
@@ -85,16 +87,12 @@ impl<'a, 'tcx> Visitor<'tcx> for TraitImplTyVisitor<'a, 'tcx> {
             // the concrete type was specified by the implementation
             if impl_ty != trait_ty {
                 if let Some(impl_ty) = impl_ty {
-                    if self.item_type == impl_ty {
-                        let is_self_ty = if let def::Res::SelfTy(..) = path.res {
-                            true
-                        } else {
-                            false
-                        };
-
-                        if !is_self_ty {
-                            span_use_self_lint(self.cx, path);
-                        }
+                    let is_self_ty = match path.res {
+                        def::Res::SelfTy(..) => true,
+                        _ => false
+                    };
+                    if self.item_type == impl_ty && !is_self_ty {
```

Comment:
This could be restructured in a `if_chain!`:
```rust
let lint = if_chain! {
    if impl_ty != trait_ty;
    if let Some(impl_ty) = impl_ty;
    if self.item_type == impl_ty;
    then {
        match path.res {
            def::Res::SelfTy(..) => {},
            _ => span_use_self_lint(self.cx, path, None),
        }
    }
};
```

---
unrelated lint idea:
```rust
let x = if [let]? _ {
    {true,false}
} else {
    {false,true}
}

if [!]?x {
    do_something();
}
```
Can be rewritten as
```rust
match _ {
    Some_::Thing(..) => do_something();
    _ => {}
}
// or 
if _ {
    do_something();
}
```

Reply:
Fixed. And I rolled in one more `if` into it.